src/wallet/wallet.cpp

```diff
@@ -802,143 +811,110 @@ void CWalletTx::GetAmounts2(list<COutputEntry>& listReceived,
-                output.amount += -nFee;
-                nFee=0;
-                if (output.amount != 0)
-                {
-                        listReceived.push_back(output);
-                        break;
-                }
-            }
-            else
+            // You can't simply use nCredit here, because we specifically are counting ALL outputs,
+            // regardless of whether they are mine or not. This is because instead of doing the coinstake
+            // as a single "net" entry, we show the whole coinstake AS IF the entire coinstake were back
+            // to the staker, and then create separate "send" entries for the sidestakes out to another
+            // address that is not mine.
+            amount += txout.nValue;
+
+            // If we are on the last output of the coinstake, then push the net amount.
+            if (i == vout.size() - 1)
             {
+                // We want the destination for the overall coinstake to come from output one,
+                // which also matches the input.
+                ExtractDestination(vout[1].scriptPubKey, address);
+
+                // For the rolled up coinstake entry, the first output is indicated in the pushed output
+                output = {address, amount - nDebit, 1};
                 listReceived.push_back(output);
             }
         }
-    }
-
-}
-
-
-
-void CWalletTx::GetAmounts(list<pair<CTxDestination, int64_t> >& listReceived,
-                           list<pair<CTxDestination, int64_t> >& listSent, int64_t& nFee, string& strSentAccount, const isminefilter& filter) const
-{
-    nFee = 0;
-    listReceived.clear();
-    listSent.clear();
-    strSentAccount = strFromAccount;
-
-    // Compute fee:
-    int64_t nDebit = GetDebit();
-    if (nDebit > 0) // debit>0 means we signed/sent this transaction
-    {
-        int64_t nValueOut = GetValueOut();
-        nFee = nDebit - nValueOut;
-    }
-
-    // Sent/received.
-    for (auto const& txout : vout)
-    {
-        // Skip special stake out
-        if (txout.scriptPubKey.empty())
-            continue;
 
-        isminetype fIsMine;
-        // Only need to handle txouts if AT LEAST one of these is true:
-        //   1) they debit from us (sent)
-        //   2) the output is to us (received)
-        if (nDebit > 0)
+        // If this is a coinstake, we are receiving the output (fIsMine != ISMINE_NO), and vout[1] is not mine,
+        // which means this is a sidestake TO me from an address that is not mine, add it as a "received" entry.
+        //
+        // If this is not a coinstake, and the output is mine, include it as a "received" entry. Note this
+        // excludes coinstake receives where the coinstake is mine, because that is handled above as a grouped entry.
+        if ((IsCoinStake() && fIsMine != ISMINE_NO && pwallet->IsMine(vout[1]) == ISMINE_NO)
+                || (!IsCoinStake() && fIsMine != ISMINE_NO))
```

Comment:
```suggestion
        if (((pwallet->IsMine(vout[1]) == ISMINE_NO) || !IsCoinStake()) && fIsMine != ISMINE_NO)
```

Reply:
I don't like it. It is simpler but to me obscures the logic.

Reply:
The `pwallet->IsMine(vout[1])` and `IsCoinStake()` is used a lot in this function, maybe we can save the results?

Reply:
> I don't like it. It is simpler but to me obscures the logic.

The latter is more clear to me but your call in the end.

Reply:
I changed my mind... and I like the idea of saving the results as a local...

Reply:
I made some further optimizations too.